sdk/opa.go

```diff
@@ -0,0 +1,191 @@
+	"github.com/open-policy-agent/opa/storage"
+	"github.com/open-policy-agent/opa/storage/inmem"
+)
+
+// OPA represents an instance of the policy engine.
+type OPA struct {
+	manager     *plugins.Manager
+	configBytes []byte
+	paths       []string
+	filter      loader.Filter
+	bundleMode  bool
+}
+
+// ConfigFile sets the configuration file to use on the OPA instance.
+func ConfigFile(fileName string) func(*OPA) error {
+	return func(opa *OPA) error {
+		bs, err := ioutil.ReadFile(fileName)
+		if err != nil {
+			return err
+		}
+		opa.configBytes = bs
+		return nil
+	}
+}
+
+// Config sets the configuration to use on the OPA instance.
+func Config(config string) func(*OPA) error {
+	return func(opa *OPA) error {
+		opa.configBytes = []byte(config)
+		return nil
+	}
+}
+
+// Paths sets the paths to search for policy, data files and bundles
+func Paths(paths []string) func(*OPA) error {
+	return func(opa *OPA) error {
+		opa.paths = paths
+		return nil
+	}
+}
+
+// New returns a new OPA object.
+func New(opts ...func(*OPA) error) (*OPA, error) {
+	opa := &OPA{}
+
+	for _, opt := range opts {
+		if err := opt(opa); err != nil {
+			return nil, err
+		}
+	}
+
+	id, err := uuid.New(rand.Reader)
+	if err != nil {
+		return nil, err
+	}
+
+	var pluginInitializers []func(*plugins.Manager)
+
+	if opa.paths != nil && len(opa.paths) > 0 {
+		loaded, err := initload.LoadPaths(opa.paths, opa.filter, opa.bundleMode, nil, false)
+		if err != nil {
+			return nil, fmt.Errorf("load error %v", err)
+		}
+		pluginInitializers = append(pluginInitializers, plugins.InitBundles(loaded.Bundles))
+		pluginInitializers = append(pluginInitializers, plugins.InitFiles(loaded.Files))
+	}
+
+	opa.manager, err = plugins.New(opa.configBytes, id, inmem.New(), pluginInitializers...)
+	if err != nil {
+		return nil, err
+	}
+
+	discovery, err := discovery.New(opa.manager)
+	if err != nil {
+		return nil, err
+	}
+
+	opa.manager.Register("discovery", discovery)
+
+	return opa, nil
+}
+
+// Start asynchronously starts the policy engine's plugins that download
+// policies, report status, etc.
+func (opa *OPA) Start(ctx context.Context) error {
+	return opa.manager.Start(ctx)
+}
+
+// Query sends a query to the OPA instance and returns the decision
+func (opa *OPA) Query(ctx context.Context, decisionPath string, input interface{}, opts ...func(*rego.Rego)) (interface{}, error) {
+	m := metrics.New()
+	var decisionID string
+	var decision interface{}
+
+	err := storage.Txn(ctx, opa.manager.Store, storage.TransactionParams{}, func(txn storage.Transaction) error {
+		var err error
+		decisionID, err = uuid.New(rand.Reader)
+		if err != nil {
+			return err
+		}
+
+		opts = append(opts,
+			rego.Metrics(m),
+			rego.Query(decisionPath),
+			rego.Input(input),
+			rego.Compiler(opa.manager.GetCompiler()),
+			rego.Store(opa.manager.Store),
+			rego.Transaction(txn))
+
+		query, err := rego.New(opts...).PrepareForEval(ctx)
+		if err != nil {
+			return err
+		}
+
```

Comment:
Could we store the prepared query on the OPA struct and not have to create it on every call to `Query` 🤔 . We would then pass the input on the eval call.

Reply:
I guess. I have assumed that the prepare call stored the "prepared state" internally so that the client wouldn't need to bother, but perhaps that was wrong. Need to check.

Reply:
@ashutosh-narkar see https://github.com/open-policy-agent/opa/pull/3186/commits/5b96a5db9c34348e5b47810f5e4006a816d66188 for an attempt at doing this without fixating the cache to a single query. Not sure how that works in the envoy plugin.. is it only ever a single decision path that's queried there?

Reply:
>  is it only ever a single decision path that's queried there?

Yes! It's serving all incoming CheckRequests based on one (configurable) query.

Reply:
@srenatus right, thanks. I don't think that's a serviceable model for an SDK function though.. especially since it would fail silently on a new query and the caller would have no idea why. Hence the use of maps keyed on the query in the last commit.